flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java

```diff
@@ -332,112 +338,104 @@ void fail(int trackingIndex) throws Exception {
 		}
 	}
 
-	private abstract static class AbstractRandomFailureStrategy implements FailureStrategy {
+	private abstract static class AbstractOnceAfterCallNumberFailureStrategy implements FailureStrategy {
 		private static final long serialVersionUID = 1L;
 
-		private final CoinToss coin;
+		private final UUID id;
+		private final int failAfterCallNumber;
+		private transient int callCounter;
 
-		private AbstractRandomFailureStrategy(CoinToss coin) {
-			this.coin = coin;
+		private AbstractOnceAfterCallNumberFailureStrategy(int failAfterCallNumber) {
+			this.failAfterCallNumber = failAfterCallNumber;
+			id = UUID.randomUUID();
 		}
 
 		@Override
 		public boolean failOrNot(int trackingIndex) throws Exception {
-			boolean generateFailure = coin.toss();
+			callCounter++;
+			boolean generateFailure = callCounter == failAfterCallNumber;
 			if (generateFailure) {
 				fail(trackingIndex);
 			}
 			return generateFailure;
 		}
 
 		abstract void fail(int trackingIndex) throws Exception;
-	}
 
-	private static class CoinToss implements Serializable {
-		private static final long serialVersionUID = 1L;
-		private static final Random rnd = new Random();
-
-		private final int probFraction;
-		private final int probBase;
+		@Override
+		public String toString() {
+			return this.getClass().getSimpleName() + " (fail after " + failAfterCallNumber + " calls)";
+		}
 
-		private CoinToss(int probFraction, int probBase) {
-			this.probFraction = probFraction;
-			this.probBase = probBase;
+		@Override
+		public boolean equals(Object o) {
+			if (this == o) {
+				return true;
+			}
+			if (o == null || getClass() != o.getClass()) {
+				return false;
+			}
+			return Objects.equals(id, ((AbstractOnceAfterCallNumberFailureStrategy) o).id);
 		}
 
-		private boolean toss() {
-			int prob = rnd.nextInt(probBase) + 1;
-			return prob <= probFraction;
+		@Override
+		public int hashCode() {
+			return id.hashCode();
 		}
 	}
 
+	@SuppressWarnings("SynchronizationOnStaticField")
```

Comment:
instead of having this class be full of static fields, we could also turn it into a "normal" class and add a static instance in the test that we reference.

So this class would then be:
```
private static class StaticMapFailureTracker {
	private final List<AtomicInteger> mapRestarts = new ArrayList<>(MAP_NUMBER);
...
```

and on `@Before` we setup the instance.  This would make it easier to re-use and you no longer need the reset calls. You could then further simplify it by setting up the data-structures in the constructor.

Reply:
th suppression can be removed now.